### PR TITLE
Explore: Add resize to split view, with Min/Max button

### LINKED ISSUE #52106: Maximise one view to full screen on dual view
### Discussed in https://github.com/grafana/grafana/discussions/51223

<div type='discussions-op-text'>

<sup>Originally posted by **cyriltovena** June 22, 2022</sup>
I love the double view in explore, I used it very often, specially when I'm reading logs and then suddenly looking at traces.

image

One thing though that is really frustrating is that if I want to focus on a view like the trace view, my only option is to actually close the logs view. But I don't want to close it because I might still searching or exploring more traces.

My suggestion is here is to allow the same as in visual studio, which collapse vertically the view I'm not focusing see below:

https://user-images.githubusercontent.com/1053421/174961860-b6133279-e15f-45f7-92ef-79e7c97a93c4.mov

</div>

### LINKED ISSUE #52415: Explore: Make the split resizable
This is an issue version of #46373 that we closed so we could focus on reimplementing it using existing Grafana components.

Use the `SplitPaneWrapper` to make the split panels in Explore resizable. Additionally if there is time, investigate #52106

### PR DESCRIPTION
**Pull request by gelicia**
What this PR does / why we need it:
A previous PR was opened (https://github.com/grafana/grafana/pull/53012) that allowed split pane resizing in Explore. After chatting about it with @Elfo404 , he had the idea to not keep track of pane width in state that would trigger a lot of renders, but instead to just keep track of width within the component. We still require some use of state changes to keep track of various sizing button pushes, but this is much more efficient. Thanks to @Elfo404 for the idea! 

- Adds a new component, `SplitView` that is a wrapper with styling to the `react-split-pane` library. This is also a change that will affect dashboard's edit page. This is so we can remove the horizontal split and other dashboard specific features and reuse it both by dashboards and explore.
- Adds +/- button to allow for maximizing a certain pane
- The sizing defaults here are kind of generous, and I'd love recommendations regarding them and the toolbar. I left the min-width of explore as 600px with a horizontal scrollbar kicking in if split. Maybe we should go down here? I was looking to avoid weird squished query boxes but 600px is most likely overcompensating. We could also look into styling the toolbar so it does not stay to the right and you have to scroll to access it.
-  There are three state variables that reflect the state of the panes. The state of the panes can be in three "modes" - either one is maximized, they are both even, or there was some manual resize where the panes are even (+- 100 pixels), or one is larger. We keep track of that in the explore state to handle what should happen when the max/min buttons are clicked, and what kind of button should show.
- I did ultimately end up needing to keep the width in the `Wrapper` component. Before this was kept, there was a bug where if the panes were manually resized, the second part of the `widthCalc` statement would fire and the variable would be set to even (it would still show up as uneven). If you then clicked the - button to make the panes even, the `widthCalc` function would be unchanged and it would not fire. By being sure the `widthCalc` result matches the resize, even if the resize is manual, we can be sure that we are changing it with every resize and avoid this. This does not impact any overall Explore state, so performance impacts should be low.

Which issue(s) this PR fixes:
Fixes https://github.com/grafana/grafana/issues/52415
Fixes https://github.com/grafana/grafana/issues/52106

### CODE REVIEW

File: public/app/features/explore/ExploreToolbar.tsx [hunk 1]
```diff
@@ -119,6 +120,16 @@ class UnConnectedExploreToolbar extends PureComponent<Props> {
       contextSrv.hasAccess(AccessControlAction.DashboardsCreate, contextSrv.isEditor) ||
       contextSrv.hasAccess(AccessControlAction.DashboardsWrite, contextSrv.isEditor);
 
+    const isLargerExploreId = largerExploreId === exploreId;
+
+    const onClickResize = () => {
+      if (isLargerExploreId) {
+        this.props.evenResizePane(true);
```
Comment (Elfo404) [inline public/app/features/explore/ExploreToolbar.tsx:127]: AFAICT this is the only place where this is used and the parameter doesn't seem to do anything, is it a leftover?

Comment (gelicia) [reply public/app/features/explore/ExploreToolbar.tsx:127]: Nope. If you hit the resize button (either + or -) and the panel is larger, it will resize the panes to be even. The prop is calling a state function to do that.

Comment (Elfo404) [reply public/app/features/explore/ExploreToolbar.tsx:127]: but it's called only here and only with `true`, so is it passing `true` really necessary?

Comment (gelicia) [reply public/app/features/explore/ExploreToolbar.tsx]: Good catch! Fixed

File: public/app/features/explore/Wrapper.tsx [hunk 2]
```diff
@@ -86,22 +118,56 @@ class WrapperUnconnected extends PureComponent<Props> {
...
+  updateSplitSize(rightPaneWidth: number) {
+    const evenSplitWidth = window.innerWidth / 2;
+    const areBothSimilar = inRange(rightPaneWidth, evenSplitWidth - 100, evenSplitWidth + 100);
+    if (areBothSimilar) {
+      this.props.changeLargerPane(undefined);
+    } else {
+      this.props.changeLargerPane(rightPaneWidth > evenSplitWidth ? ExploreId.right : ExploreId.left);
+    }
+
+    this.setState({ rightPaneWidth });
```
Comment (gelicia) [inline public/app/features/explore/Wrapper.tsx:130]: We did end up needing to keep the width in component state here. This was because a manual resize would look fine fine, but the `widthCalc` would remain constant, so trying to do a minimize/maximize after a manual resize wouldn't work. We need the widthCalc to reflect every width change.

To see what I'm talking about, try a commit before this one. Manually drag the bar to one side, and hit - on the larger pane. It won't work until you click it twice more.

File: public/app/core/components/SplitPaneWrapper/SplitView.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    width: ${theme.spacing(2)};
+  `;
+
+export const SplitView = ({ uiState: { rightPaneSize }, children, minSize = 200, onResize }: Props) => {
+  const { width } = useViewportSize();
+
+  // create two elements for library, even if only one exists (one will be hidden)
+  const hasSplit = children.filter(Boolean).length === 2;
+
+  const existingChildren = [
```
Comment (gelicia) [inline public/app/core/components/SplitPaneWrapper/SplitView.tsx:73]: @Elfo404 This is how I propose getting around the library limitation. This is an addition to the edits you provided. This allows us to key the panes so React doesn't complain :)

**Comment by gelicia** [comment]
The failing e2e test involves the virtualized list involved in jaeger trace views. I confirmed that it is in fact rendering the whole list and not doing redraws as expected on scroll. I will be working with the traces team on how to fix this.

File: public/app/features/explore/ExploreToolbar.tsx [hunk 4]
```diff
@@ -119,6 +120,16 @@ class UnConnectedExploreToolbar extends PureComponent<Props> {
       contextSrv.hasAccess(AccessControlAction.DashboardsCreate, contextSrv.isEditor) ||
       contextSrv.hasAccess(AccessControlAction.DashboardsWrite, contextSrv.isEditor);
 
+    const isLargerExploreId = largerExploreId === exploreId;
```
Comment (ifrost) [inline public/app/features/explore/ExploreToolbar.tsx:123]: I know we do it in many places, but I think `ExploreToolbar` shouldn't know if it's larger than the other pane (ideally shouldn't need to know `exploreId` at all). Could we instead pass `mode` down to the toolbar and decide what interactions are available based on the `mode`?

File: public/app/features/explore/ExploreToolbar.tsx [hunk 5]
```diff
@@ -153,9 +164,14 @@ class UnConnectedExploreToolbar extends PureComponent<Props> {
                 Split
               </ToolbarButton>
             ) : (
-              <ToolbarButton tooltip="Close split pane" onClick={() => closeSplit(exploreId)} icon="times">
-                Close
-              </ToolbarButton>
+              <>
+                <ToolbarButton title="Resize pane" icon="columns" disabled={isLive} onClick={onClickResize}>
```
Comment (ifrost) [inline public/app/features/explore/ExploreToolbar.tsx:168]: The icon looks quite off compared to other icons in the toolbar, also the tooltip doesn't tell what type of resize it is. Could we use simpler icons that would represent maximizing/minimizing? At the moment it looks more like it represents adding or removing panes.

Screen Shot 2022-09-12 at 14 01 26

File: public/app/features/explore/Wrapper.tsx [hunk 6]
```diff
@@ -86,22 +118,56 @@ class WrapperUnconnected extends PureComponent<Props> {
     document.title = documentTitle;
   }
 
+  updateSplitSize(rightPaneWidth: number) {
+    const evenSplitWidth = window.innerWidth / 2;
```
Comment (ifrost) [inline public/app/features/explore/Wrapper.tsx:122]: Shouldn't we use container's width, not the window width?

**Review by ifrost — COMMENTED** [review]
Works really nice! 🚀

File: public/app/features/explore/state/main.ts [hunk 7]
```diff
@@ -173,6 +201,9 @@ export const initialExploreState: ExploreState = {
   richHistoryStorageFull: false,
   richHistoryLimitExceededWarningShown: false,
   richHistoryMigrationFailed: false,
+  largerExploreId: undefined,
```
[the remark this answers is not in the document]
Comment (gelicia) [reply public/app/features/explore/state/main.ts:204]: Because initially I had implemented it with the width in redux, and then reimplemented it with only these state variables in redux so the toolbar could inteact with it and affect the pane size. I can  re-write this to put everything in Wrapper but I am curious on what that would gain us.

File: public/app/features/explore/ExploreToolbar.tsx (hunk 4 above)
Comment (gelicia) [reply public/app/features/explore/ExploreToolbar.tsx:123]: Just to be sure I understand you, your suggestion is instead of getting a `largerExploreId`, it would get something like `largerExploreMode` set as a boolean of true or false?

File: public/app/features/explore/ExploreToolbar.tsx (hunk 5 above)
Comment (gelicia) [reply public/app/features/explore/ExploreToolbar.tsx:168]: This is submitted to tomorrow's UX feedback session to discuss the icons and tooltip.

File: public/app/features/explore/Wrapper.tsx (hunk 6 above)
Comment (gelicia) [reply public/app/features/explore/Wrapper.tsx:122]: This boolean is to know what the threshold is for an even split. The `rightPaneWidth` parameter is the container's width.

File: public/app/features/explore/state/main.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/features/explore/state/main.ts:204]: >  I am curious on what that would gain us.

I know we do it quite frequently but I'm just not sure what's the advantage of pulling the state up to Redux if it's fully controlled by the Wrapper component. Was it to avoid passing these props down to ExploreToolbar?

File: public/app/features/explore/ExploreToolbar.tsx (hunk 4 above)
Comment (ifrost) [reply public/app/features/explore/ExploreToolbar.tsx:123]: I was thinking about an `enum SizeMode { NORMAL, MAXIMIZED, MINIMIZED }`? To be honest, it's a really minor thing about readability - I think it'd be just more explicit and natural. To be it feels like `ExploreToolbar` shouldn't be aware of which pane is larger, and instead, it should know if it was maximized or not by the parent container/split view.

Comment (gelicia) [reply public/app/features/explore/ExploreToolbar.tsx:123]: One of the comments from the feedback session involved changing the +/- buttons to < and > , and which one to display depends on the explore ID and size mode. I think we might not be able to avoid this, but I also think the ExploreToolbar knowing what ExploreID it has is fine.

File: public/app/features/explore/state/main.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (gelicia) [reply public/app/features/explore/state/main.ts:204]: Yeah, I guess if it's between having three variables in redux or passing three variables around the whole Wrapper, ExplorePaneContainer, Explore, ExploreToolbar I would rather just have it in redux. 

If this is important to you I can change it.

File: public/app/features/explore/ExploreToolbar.tsx (hunk 5 above)
Comment (gelicia) [reply public/app/features/explore/ExploreToolbar.tsx]: After some discussion and looking at what icons we have available, I think a simple > and < on this portrays the action nicely. I also changed the tooltip to be more informative. Let me know what you think!

File: public/app/features/explore/state/main.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ifrost) [reply public/app/features/explore/state/main.ts:204]: I think <Context> would be easier here, but you can leave it as it is. I need to look at other places that follow the same approach to see if can simplify it.

File: public/app/features/explore/Wrapper.tsx (hunk 6 above)
Comment (ifrost) [reply public/app/features/explore/Wrapper.tsx:122]: I meant it just happens that Wrapper spans the window width, but if we ever change it the threshold calculations will be incorrect. Never mind, I think it was like this before so you can leave it.

**Review by ifrost — APPROVED** [review]
Fantastic feature 🎉  It'll greatly improve logs to traces flow!

I think there are no test added for resizing. Do you think we could add some coverage?

File: public/app/features/explore/ExploreToolbar.tsx [hunk 8]
```diff
@@ -153,9 +165,20 @@ class UnConnectedExploreToolbar extends PureComponent<Props> {
...
+              <>
+                <ToolbarButton
+                  title={`${isLargerExploreId ? 'Narrow' : 'Widen'} pane`}
+                  disabled={isLive}
+                  onClick={onClickResize}
+                >
+                  {(exploreId === 'left' && isLargerExploreId) || (exploreId === 'right' && !isLargerExploreId)
+                    ? '<'
+                    : '>'}
+                </ToolbarButton>
```
Comment (Elfo404) [inline public/app/features/explore/ExploreToolbar.tsx:177]:
```suggestion
                <ToolbarButton
                  tooltip={`${isLargerExploreId ? 'Narrow' : 'Widen'} pane`}
                  disabled={isLive}
                  onClick={onClickResize}
                  icon={
                    (exploreId === 'left' && isLargerExploreId) || (exploreId === 'right' && !isLargerExploreId)
                      ? 'angle-left'
                      : 'angle-right'
                  }
                />
```

we should probably use icons instead of text for the button, also, afaik, there has been some talk around stopping using titles for buttons in favor of tooltips, which internally should also provide a label to the button

Screenshot 2022-09-20 at 11 41 00

File: public/app/features/explore/Wrapper.tsx [hunk 9]
```diff
@@ -77,7 +109,7 @@ class WrapperUnconnected extends PureComponent<Props> {
     }
   }
 
-  componentDidUpdate() {
+  componentDidUpdate(prevProps: Props) {
```
Comment (Elfo404) [inline public/app/features/explore/Wrapper.tsx:112]: `prevProps` is unused
```suggestion
  componentDidUpdate() {
```

File: public/app/features/explore/Wrapper.tsx [hunk 10]
```diff
@@ -87,22 +119,56 @@ class WrapperUnconnected extends PureComponent<Props> {
     document.title = documentTitle;
   }
 
+  updateSplitSize(rightPaneWidth: number) {
```
Comment (Elfo404) [inline public/app/features/explore/Wrapper.tsx:122]: kinda nit for readability: can we make this an arrow function so that on line 156 we can do
```
<SplitView [...] onResize={this.updateSplitSize}>[...]
```

File: public/app/features/explore/state/main.ts [hunk 11]
```diff
@@ -133,6 +143,24 @@ export function splitClose(itemId: ExploreId): ThunkResult<void> {
...
+    dispatch(maximizePaneAction({ exploreId }));
+  };
+}
+
+export function evenResizePane(): ThunkResult<void> {
+  return (dispatch, getState) => {
+    dispatch(evenPaneResizeAction());
+  };
+}
+
```
Comment (Elfo404) [inline public/app/features/explore/state/main.ts:163]: given these thunks are not doing anything async and are just proxying call to plain action creators, we should be using actions creators directly instead.

File: public/app/features/query/components/QueryEditorRow.tsx [hunk 12]
```diff
@@ -81,12 +81,11 @@ export class QueryEditorRow<TQuery extends DataQuery> extends PureComponent<Prop
...
   };
 
-  componentDidMount() {
+  async componentDidMount() {
     const { data, query } = this.props;
     const dataFilteredByRefId = filterPanelDataToQuery(data, query.refId);
     this.setState({ data: dataFilteredByRefId });
-
-    this.loadDatasource();
+    await this.loadDatasource();
```
Comment (Elfo404) [inline public/app/features/query/components/QueryEditorRow.tsx:88]: why is this change needed?

Comment (gelicia) [reply public/app/features/query/components/QueryEditorRow.tsx:91]: This was leftover from when I was trying to avoid the state change error. I'll remove it.

File: public/app/features/explore/Wrapper.test.tsx [hunk 13]
```diff
@@ -261,12 +266,36 @@ describe('Wrapper', () => {
...
 
-      store.dispatch(splitOpen<any>({ datasourceUid: 'elastic', query: { expr: 'error' } }) as any);
+      store.dispatch(mainState.splitOpen<any>({ datasourceUid: 'elastic', query: { expr: 'error' } }) as any);
 
       // Editor renders the new query
       await screen.findByText(`elastic Editor input: error`);
       await screen.findByText(`loki Editor input: { label="value"}`);
     });
+
+    it('handles split size events and sets relevant variables', async () => {
```
Comment (gelicia) [inline public/app/features/explore/Wrapper.test.tsx:276]: @ifrost @Elfo404 Here is the test I arrived to. I wanted to test pane width but it doesn't seem to be easy to do - the element that gets a CSS width is the autosizer, which is mocked in this to always return a div of the same size. All the other elements I can select return 0 as the width.

File: public/app/features/explore/Wrapper.test.tsx [hunk 14]
```diff
@@ -261,12 +266,36 @@ describe('Wrapper', () => {
...
+      expect(narrowButton.length).toBe(0);
+      expect(Number.parseInt(getComputedStyle(panes[0]).width, 10)).toBe(1000);
+      expect(Number.parseInt(getComputedStyle(panes[1]).width, 10)).toBe(1000);
+      const resizer = screen.getByRole('presentation');
+      fireEvent.mouseDown(resizer, { buttons: 1 });
+      fireEvent.mouseMove(resizer, { clientX: -700, buttons: 1 });
+      fireEvent.mouseUp(resizer);
+      widenButton = await screen.findAllByLabelText('Widen pane');
+      narrowButton = await screen.queryAllByLabelText('Narrow pane');
+      panes = screen.getAllByRole('main');
```
Comment (ifrost) [inline public/app/features/explore/Wrapper.test.tsx:294]: Is it needed?

**Review by ifrost — APPROVED** [review]
Great work 🎉

File: public/app/features/explore/Wrapper.test.tsx (hunk 14 above)
Comment (gelicia) [reply public/app/features/explore/Wrapper.test.tsx:294]: oops, nope. Leftover from trying to find a place to get the width before and after.

File: public/app/core/components/SplitPaneWrapper/SplitView.tsx (hunk 3 above)
Comment (torkelo) [reply public/app/core/components/SplitPaneWrapper/SplitView.tsx:73]: @gelicia this is a little bit bad refactor as the resizer styles are now duplicated in SplitPaneWrapper and SplitView (and causing some bugs that where hard to track down, the horizontal reizer has a border now)

Comment (gelicia) [reply public/app/core/components/SplitPaneWrapper/SplitView.tsx:73]: @torkelo We did duplicate the styles yes, but the reasoning was that the SplitPaneWrapper had a lot of logic in it that was specific to dashboards, and we wanted another layer of abstraction that would keep things like styles that would be used universally in Grafana but without the horizontal and vertical splits that are used in dashboards. I realize now I should have brought the dashboards team in to review this directly. My apologies.

Comment (torkelo) [reply public/app/core/components/SplitPaneWrapper/SplitView.tsx:73]: @gelicia no worries

Comment (torkelo) [reply public/app/core/components/SplitPaneWrapper/SplitView.tsx:73]: > wanted another layer of abstraction

But the horizontal split was still left in SplitPaneWrapper, so not really another level of abstraction, just a new duplicated abstraction for vertical splits, one that now has a different (and duplicated) design and behaviors (And missing features for persistance) 

https://github.com/grafana/grafana/pull/56190
